### DISCUSSION TITLE
Does SQLAlchemy support using stored procedures for CRUD operations?

### DISCUSSION THREAD

**Discussion by gotdang** [sqlalchemy/sqlalchemy#9714] [Usage Questions] [answered]
I'm working in an environment where all database operations, including CRUD, are performed via stored procedures. Is there a way to ask SQLAlchemy to use these stored procedures instead of SQL statements? I'm not asking if SA can call arbitrary stored procedures; I know it can do that. Can I perform CRUD operations on a server where the connection can't execute arbitrary SQL statements, only execute procedures?

Below is some minimal sample (and functioning) code on a DB where I can execute SQL statements.
I'd like this code to work more-or-less untouched but in a sproc-only world.

"Obviously", I'd have to write the SQL statement calls; I don't expect SA to magically know the appropriate procedures and parameters. I'm asking if there is a hook for this, where I can intercept the SQL and replace it with a stored procedure call.

Thank you,
    --dang
```python
# Conventional imports for a SQLAlchemy project have been elided.
class Base(DeclarativeBase):
    ...

class Account(Base):
    __tablename__ = 'account'
    account_id = mapped_column(Integer, primary_key=True, nullable=False, server_default=FetchedValue())
    account_nm = mapped_column(String(50), nullable=False)

session = Session(engine)

a1 = Account(account_nm='First account')
a2 = Account(account_nm='Second account')

session.add(a1)
session.add(a2)
session.commit()

a1.account_nm = 'First account (updated)'
session.add(a1)
session.commit()

accounts = session.query(Account)
for account in accounts:
    print(account)

session.delete(a2)
session.delete(a1)
session.commit()
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

Probably mike can with it more on thins, but short answer is not really.

A longer answer is that you could use the [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event to rewrite the statement from what sqlalchemy is trying to run to the stored procedure call. That said I don't expect it to be easy

For select only statement is easier since you would use this https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#orm-queryguide-selecting-text, but I'm not aware of a similar thing for insert/update/delete

**Reply by gotdang** [reply]
Thank you. I'll look into do_orm_execute more; it looks like a reasonable place to start.

**Reply by CaselIT** [reply]
You may also need the other event hooks for the session since not all queries may pass for that event (iirc some on the unit of work may use other events, like https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_insert )

**Comment by zzzeek** [comment]
unit of work calling into SPs?  no that's not possible.

**Reply by gotdang** [reply]
If the session object (or whatever the relevant object would be) passes the entire transaction, perhaps as json or XML, not just a single row's data, it's very possible.
